visualization/app/codeCharta/ui/codeMap/codeMap.label.service.spec.ts

```diff
@@ -0,0 +1,197 @@
+            return {
+                camera: {
+                    position: {
+                        distanceTo: jest.fn()
+                    }
+                }
+            }
+        })();
+    }
+
+    function withMockedThreeSceneService() {
+        services.threeSceneService = jest.fn<ThreeSceneService>(()=>{
+            return {
+                mapGeometry: jest.fn(),
+                labels: {
+                    add: jest.fn(),
+                    children: jest.fn()
+                }
+            }
+        })();
+    }
+
+    function setCanvasRenderSettings() {
+
+        sampleRenderSettings = {
+            heightKey: "a",
+            colorKey: "b",
+            renderDeltas: false,
+            colorRange: {
+                from: 1,
+                to: 2,
+                flipped: false
+            },
+            mapSize: 500,
+            deltaColorFlipped: false
+        };
+
+        sampleLeaf = {
+            name: "sample",
+            width: 1,
+            height: 2,
+            length: 3,
+            depth: 4,
+            x0: 5,
+            z0: 6,
+            y0: 7,
+            isLeaf: true,
+            deltas: {"a":1, "b":2},
+            attributes: {"a":20, "b":15},
+            children: [],
+            isDelta: false
+        };
+
+        canvasCtxMock = {
+            font: "",
+            measureText: jest.fn(),
+            fillRect: jest.fn(),
+            fillText: jest.fn(),
+            strokeRect: jest.fn()
+        };
+
+        createElementOrigin = document.createElement;
+
+        document.createElement = ()=>{
+            return {
+                getContext: ()=>{
+                    return canvasCtxMock;
+                }
+            };
+        };
+
+        canvasCtxMock.measureText.mockReturnValue({width: 10});
+
+    }
+
+    afterEach(()=>{
+        document.createElement = createElementOrigin;
+    });
+
+    it("should have no labels stored after construction", ()=>{
+        expect(codeMapLabelService.labels.length).toBe(0);
+    });
+
+    it("addLabel should add label if node has a height attribute mentioned in renderSettings", ()=>{
+        codeMapLabelService.addLabel(sampleLeaf, sampleRenderSettings);
+        expect(codeMapLabelService.labels.length).toBe(1);
+    });
+
+    it("addLabel should not add label if node has not a height attribute mentioned in renderSettings", ()=>{
+        sampleLeaf.attributes = {"notsome": 0};
+        sampleRenderSettings.heightKey = "some";
+        codeMapLabelService.addLabel(sampleLeaf, sampleRenderSettings);
+        expect(codeMapLabelService.labels.length).toBe(0);
+    });
+
+    it("addLabel should calculate correct height without delta", ()=>{
+        codeMapLabelService.addLabel(sampleLeaf, sampleRenderSettings);
+        let positionWithoutDelta: Vector3 = codeMapLabelService.labels[0].sprite.position;
+        expect(positionWithoutDelta.y).toBe(88);
```

Comment:
Does anybody know where the number `88` comes from? The labels seem to work in the visualisation, but the test now results in `failed` with the number `93`...

Reply:
Okey, I found the reason: The value changed, because I changed the margin of the label in [codeMap.label.service.ts#L89](https://github.com/MaibornWolff/codecharta/blob/feature/237/high-margin-kills-label-size/visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts#L89).